examples/file-upload.html

```diff
@@ -0,0 +1,181 @@
+  <h2>File Upload</h2>
+</hgroup>
+
+<file-upload>
+
+  <input multiple id=files name=files[] type=file>
+  <label for=files ondragover="event.preventDefault()" ondragenter=ondragstart ondragleave=ondragend ondrop=ondragdrop>
+    Click to upload file...
+  </label>
+
+  <table style="width:300px">
+    <thead style="text-align: left">
+      <tr>
+        <th width="1%"></th>
+        <th>Name</th>
+        <th width="10%">Size</th>
+        <th width="1%"></th>
+      </tr>
+    </thead>
+    <tbody>
+      <template name=files>
+        <tr>
+
+          <td>
+            <img src={icon_url} alt={type} hidden={hidden} style="width: 1em; height: 1em">
+          </td>
+          <td>
+            {name}
+          </td>
+          <td>
+            {size}
+          </td>
+          <td style="cursor:pointer">
+            <a onclick=onremove data-index={#}>&times;</a>
+          </td>
+        </tr>
+      </template>
+    </tbody>
+  </table>
+
+  <a onclick=onclear>Clear</a>
+
+<style>
+
+file-upload input[type=file] {
+  opacity: 0;
+  z-index: -1;
+  width: 0.1px;
+  height: 0.1px;
+  overflow: hidden;
+  position: absolute
+}
+file-upload input[type=file] + label {
+  display: block;
+  padding: 2em 0;
+  margin: 1em 20%;
+  cursor: pointer;
+  font-size: 1.25em;
+  text-align: center;
+  color: rgba(20,20,20,0.6);
+  background-color: rgba(240,240,240, 0.8)
+}
+
+file-upload input[type=file] + label:hover
+  { background-color: rgba(240,240,240, 1.0) }
+
+</style>
+
+<script>
+
+Element `file-upload`
+
+(class extends HTMLElement {
+
+  initialize ()
+    { this.context.files = [] }
+
+
+  static onchange (event) {
+    if (!!! event.target.files) return
+
+    this.add_files(event.target.files)
+
+    this.render ()
+  }
+
+  static onremove (event) {
+    this.context.files.splice(event.target.dataset.index, 1)
+
+    this.render ()
+  }
+
+  static onclear (event) {
+    this.context.files = []
+
+    this.render ()
+  }
+
+  static ondragstart (event) {
+    event.target.innerHTML = 'Drop here!'
+
+    event.preventDefault ()
+  }
+
+  static ondragend (event) {
+    this.reset_label(event.target)
+  }
+
+  static ondragdrop (event) {
+    this.reset_label(event.target)
+
+    event.preventDefault ()
+
+    var files = event.dataTransfer.files
+
+    if (!!! files) return
+
+    this.add_files(files)
+
+    this.render ()
+  }
+
+  reset_label (label) {
+    label.innerHTML = 'Click to upload file...'
+  }
+
+  add_files (files) {
+    // Should have a sanity check.
+    // All `files` should be of class File
+    this.context.files =
+      this.context.files.concat (...files)
+  }
+
+  get files () {
+    return this.context.files
+      .map ((file, idx) => this.constructor.add_icon_url(file)) 
+  }
+
+
+  static add_icon_url (file) {
+    var icon_url = this.mimetype_icons[file.type]
+
+    file.icon_url = icon_url === undefined ? '' : icon_url
+    file.hidden_icon = icon_url === undefined ? 'hidden' : ''
+
+    return file
```

Comment:
@VicenteRD methinks you can get away with removing `return file` I think `map ()` returns a duped object.